File: hello-grpc-java/src/main/java/org/feuyeux/grpc/common/ErrorMapper.java
package org.feuyeux.grpc.common;

import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import java.util.HashMap;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;

/**
 * Error mapper for translating gRPC status codes to human-readable messages and determining whether
 * an error is retryable.
 */
@Slf4j
public class ErrorMapper {

  /**
   * Maps gRPC status codes to human-readable error messages
   *
   * @param throwable The exception to map
   * @return Human-readable error message
   */
  public static String mapGrpcError(Throwable throwable) {
    if (throwable == null) {
      return "Success";
    }

    Status status;
    String message;

    if (throwable instanceof StatusRuntimeException sre) {
      status = sre.getStatus();
      message = status.getDescription() != null ? status.getDescription() : "";
    } else {
      return "Unknown error: " + throwable.getMessage();
    }

    String description = getStatusDescription(status.getCode());

    if (!message.isEmpty()) {
      return description + ": " + message;
    }
    return description;
  }

  /**
   * Gets a human-readable description for a gRPC status code
   *
   * @param code The status code
   * @return Human-readable description
   */
  private static String getStatusDescription(Status.Code code) {
    return switch (code) {
      case OK -> "Success";
      case CANCELLED -> "Operation cancelled";
      case UNKNOWN -> "Unknown error";
      case INVALID_ARGUMENT -> "Invalid request parameters";
      case DEADLINE_EXCEEDED -> "Request timeout";
      case NOT_FOUND -> "Resource not found";
      case ALREADY_EXISTS -> "Resource already exists";
      case PERMISSION_DENIED -> "Permission denied";
      case RESOURCE_EXHAUSTED -> "Resource exhausted";
      case FAILED_PRECONDITION -> "Precondition failed";
      case ABORTED -> "Operation aborted";
      case OUT_OF_RANGE -> "Out of range";
      case UNIMPLEMENTED -> "Not implemented";
      case INTERNAL -> "Internal server error";
      case UNAVAILABLE -> "Service unavailable";
      case DATA_LOSS -> "Data loss";
      case UNAUTHENTICATED -> "Authentication required";
    };
  }

  /**
   * Determines if an error should be retried
   *
   * @param throwable The exception to check
   * @return true if the error is retryable, false otherwise
   */
  public static boolean isRetryableError(Throwable throwable) {
    if (throwable == null) {
      return false;
    }

    if (!(throwable instanceof StatusRuntimeException sre)) {
      return false;
    }

    Status.Code code = sre.getStatus().getCode();

    return switch (code) {
      case UNAVAILABLE, DEADLINE_EXCEEDED, RESOURCE_EXHAUSTED, INTERNAL -> true;
      default -> false;
    };
  }

  /**
   * Handles RPC errors with logging and context
   *
   * @param throwable The exception that occurred
   * @param operation The operation name
   * @param context Additional context information
   */
  public static void handleRpcError(
      Throwable throwable, String operation, Map<String, Object> context) {
    if (throwable == null) {
      return;
    }

    String errorMsg = mapGrpcError(throwable);
    Map<String, Object> logContext = new HashMap<>(context);
    logContext.put("operation", operation);
    logContext.put("error", errorMsg);

    if (isRetryableError(throwable)) {
      log.warn("Retryable error occurred: {}", logContext);
    } else {
      log.error("Non-retryable error occurred: {}", logContext);
    }
  }
}

File: hello-grpc-csharp/Common/ErrorMapper.cs
using System;
using System.Collections.Generic;
using Grpc.Core;
using log4net;

namespace Common
{
    /// <summary>
    /// Error mapper for translating gRPC status codes to human-readable messages
    /// and determining whether an error is retryable.
    /// </summary>
    public static class ErrorMapper
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(ErrorMapper));

        /// <summary>
        /// Maps gRPC RpcException to human-readable error message
        /// </summary>
        /// <param name="exception">The exception to map</param>
        /// <returns>Human-readable error message</returns>
        public static string MapGrpcError(Exception exception)
        {
            if (exception == null)
            {
                return "Success";
            }

            if (exception is RpcException rpcException)
            {
                string description = GetStatusDescription(rpcException.StatusCode);
                string message = rpcException.Status.Detail;

                if (!string.IsNullOrEmpty(message))
                {
                    return $"{description}: {message}";
                }
                return description;
            }

            return $"Unknown error: {exception.Message}";
        }

        /// <summary>
        /// Gets human-readable description for a gRPC status code
        /// </summary>
        /// <param name="code">The status code</param>
        /// <returns>Human-readable description</returns>
        private static string GetStatusDescription(StatusCode code)
        {
            return code switch
            {
                StatusCode.OK => "Success",
                StatusCode.Cancelled => "Operation cancelled",
                StatusCode.Unknown => "Unknown error",
                StatusCode.InvalidArgument => "Invalid request parameters",
                StatusCode.DeadlineExceeded => "Request timeout",
                StatusCode.NotFound => "Resource not found",
                StatusCode.AlreadyExists => "Resource already exists",
                StatusCode.PermissionDenied => "Permission denied",
                StatusCode.ResourceExhausted => "Resource exhausted",
                StatusCode.FailedPrecondition => "Precondition failed",
                StatusCode.Aborted => "Operation aborted",
                StatusCode.OutOfRange => "Out of range",
                StatusCode.Unimplemented => "Not implemented",
                StatusCode.Internal => "Internal server error",
                StatusCode.Unavailable => "Service unavailable",
                StatusCode.DataLoss => "Data loss",
                StatusCode.Unauthenticated => "Authentication required",
                _ => "Unknown error code"
            };
        }

        /// <summary>
        /// Determines if an error should be retried
        /// </summary>
        /// <param name="exception">The exception to check</param>
        /// <returns>true if the error is retryable, false otherwise</returns>
        public static bool IsRetryableError(Exception exception)
        {
            if (exception == null)
            {
                return false;
            }

            if (exception is not RpcException rpcException)
            {
                return false;
            }

            return rpcException.StatusCode switch
            {
                StatusCode.Unavailable or 
                StatusCode.DeadlineExceeded or 
                StatusCode.ResourceExhausted or 
                StatusCode.Internal => true,
                _ => false
            };
        }

        /// <summary>
        /// Logs an error with request ID and operation context
        /// </summary>
        /// <param name="exception">The exception that occurred</param>
        /// <param name="requestId">The request ID</param>
        /// <param name="operation">The operation name</param>
        public static void LogError(Exception exception, string requestId, string operation)
        {
            if (exception == null)
            {
                return;
            }

            string errorMsg = MapGrpcError(exception);
            Log.Error($"[{operation}] Request {requestId} failed: {errorMsg}");
        }

        /// <summary>
        /// Handles RPC errors with logging and context
        /// </summary>
        /// <param name="exception">The exception that occurred</param>
        /// <param name="operation">The operation name</param>
        /// <param name="context">Additional context information</param>
        public static void HandleRpcError(Exception exception, string operation, Dictionary<string, object> context = null)
        {
            if (exception == null)
            {
                return;
            }

            string errorMsg = MapGrpcError(exception);
            var logContext = new Dictionary<string, object>(context ?? new Dictionary<string, object>())
            {
                ["operation"] = operation,
                ["error"] = errorMsg
            };

            string contextStr = string.Join(", ", logContext);

            if (IsRetryableError(exception))
            {
                Log.Warn($"Retryable error occurred: {contextStr}");
            }
            else
            {
                Log.Error($"Non-retryable error occurred: {contextStr}");
            }
        }
    }
}
